cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/PredefinedAbstainDRep.hs

```diff
@@ -0,0 +1,334 @@
+                      , "--prev-governance-action-index", show prevGovernanceActionIndex
+                      ]) prevGovActionInfo ++
+    [ "--number-of-pools", show desiredPoolNumber
+    , "--anchor-url", "https://tinyurl.com/3wrwb2as"
+    , "--anchor-data-hash", proposalAnchorDataHash
+    , "--out-file", proposalFile
+    ]
+
+  proposalBody <- H.note $ baseDir </> "tx.body"
+  txIn <- findLargestUtxoForPaymentKey epochStateView sbe wallet
+
+  void $ H.execCli' execConfig
+    [ "conway", "transaction", "build"
+    , "--change-address", Text.unpack $ paymentKeyInfoAddr wallet
+    , "--tx-in", Text.unpack $ renderTxIn txIn
+    , "--proposal-file", proposalFile
+    , "--out-file", proposalBody
+    ]
+
+  signedProposalTx <- signTx execConfig cEra baseDir "signed-proposal"
+                             (File proposalBody) [paymentKeyInfoPair wallet]
+
+  submitTx execConfig cEra signedProposalTx
+
+  governanceActionTxId <- retrieveTransactionId execConfig signedProposalTx
+
+  !propSubmittedResult <- findCondition (maybeExtractGovernanceActionIndex sbe (fromString governanceActionTxId))
+                                        (unFile configurationFile)
+                                        (unFile socketPath)
+                                        (EpochNo 30)
+
+  governanceActionIndex <- case propSubmittedResult of
+                             Left e ->
+                               H.failMessage callStack
+                                 $ "findCondition failed with: " <> displayError e
+                             Right Nothing ->
+                               H.failMessage callStack "Couldn't find proposal."
+                             Right (Just a) -> return a
+
+  return (governanceActionTxId, governanceActionIndex)
+
+voteChangeProposal :: (MonadTest m, MonadIO m, MonadCatch m, H.MonadAssertion m)
+  => H.ExecConfig
+  -> EpochStateView
+  -> ShelleyBasedEra ConwayEra
+  -> FilePath
+  -> FilePath
+  -> String
+  -> Word32
+  -> [([Char], Int)]
+  -> PaymentKeyInfo
+  -> m ()
+voteChangeProposal execConfig epochStateView sbe work prefix governanceActionTxId governanceActionIndex votes wallet = do
+  baseDir <- H.createDirectoryIfMissing $ work </> prefix
+
+  let era = toCardanoEra sbe
+      cEra = AnyCardanoEra era
+
+  voteFiles <- generateVoteFiles execConfig baseDir "vote-files"
+                                 governanceActionTxId governanceActionIndex
+                                 [(defaultDRepKeyPair idx, vote) | (vote, idx) <- votes]
+
+  voteTxBodyFp <- createVotingTxBody execConfig epochStateView sbe baseDir "vote-tx-body"
+                                     voteFiles wallet
+
+  voteTxFp <- signTx execConfig cEra baseDir "signed-vote-tx" voteTxBodyFp
+                     (paymentKeyInfoPair wallet:[defaultDRepKeyPair n | (_, n) <- votes])
+  submitTx execConfig cEra voteTxFp
+
+getDesiredPoolNumberValue :: (MonadTest m, MonadCatch m, MonadIO m) => H.ExecConfig -> m Integer
```

Comment:
Elaborate on what the "desired pool number value is". I don't want to have to go to `cardano-ledger` to figure it out and I want to avoid others from making the same trip. Something like "the protocol parameter that defines the optimal number of SPOs" would be sufficient.


Get in the habit of looking at the shelley spec and adding comments to code we write here to make it easier on the reader.

Reply:
Fixed in: https://github.com/IntersectMBO/cardano-node/commit/de066556bb3c18a94fdb72e2e57c703016783886